### PR TITLE
feat!: implementing `hash.Hash` interface

### PR DESCRIPTION
**Pull request by distractedm1nd**
## Overview

Resolves a TODO from celestia-node:
https://github.com/celestiaorg/celestia-node/blob/8f13efab8aa314d52708dea8873eea70027745cb/share/ipld/namespace_hasher.go#L21-L33

Motivation: Instead of only providing utilities that require knowing if a value is an inner node or a leaf to hash, implement the `hash.Hash` interface so that `Write`/`Sum` can be used generally over both node types.

This PR is breaking as it changes the `Hasher` constructor to include share size.

To be discussed:
- [ ] Is this even desired? We want it in node but hasn't been discussed elsewhere AFAIK
- [ ] This PR introduces testify a dep, I can rewrite the tests to exclude this if unwanted
- [ ] Why is NamespaceLen exported, and has a getter in addition? Should `shareSize` also be exported then?

## Checklist

- [X] New and updated code has appropriate documentation
- [X] New and updated code has new and/or updated testing
- [ ] Required CI checks are passing

### CODE REVIEW

**Review by liamsi — COMMENTED** [review]
Looks great! Will do another pass tomorrow morning. 
Linking this issue as it is somewhat related: https://github.com/celestiaorg/nmt/issues/87

File: nmt.go [hunk 1]
```diff
@@ -11,6 +11,12 @@ import (
 	"github.com/celestiaorg/nmt/namespace"
 )
 
+const (
+	DefaultNamespaceIDLen = 8
+	DefaultShareSize      = 512
```
Comment (Wondertan) [inline nmt.go:16]: Might be worth calling it LeafSiae as Share is a concept outside of the NMT and there is a notion of Leaf already in NMT.

File: hasher.go [hunk 2]
```diff
@@ -69,18 +74,76 @@ func (n *Hasher) NamespaceSize() namespace.IDSize {
...
+
+	ln := len(data)
+	switch ln {
+	default:
+		return 0, fmt.Errorf("invalid data size written to the hasher, len: %v", ln)
+	// inner nodes are made up of the nmt hashes of the left and right children
+	case n.Size() * 2:
+		n.tp = NodePrefix
+	// leaf nodes contain the namespace length and a share
+	case int(n.NamespaceLen) + n.shareSize:
```
Comment (Wondertan) [inline hasher.go:110]: We might consider making leaf share a default case, then we wouldn't need to pass shareSize. 

It's ok to unlimit the share/leaf size as NMT can be used outside of the celestia'a use case, where leafs could have different sizes.

There is no security issues with this as integrity is checked anyway. As I can see so far no other panics would be introduced, but we should triple check that if decide leaf to be a default case.

**Comment by Wondertan** [comment]
Doesn't diverge from implementation in celestia-node so LGTM

**Review by liamsi — COMMENTED** [review]
Agree with @Wondertan's comments (particularly using LeafSize instead of ShareSize but also not enforcing a fixed share size here). Otherwise LGTM 👏🏼

**Review by Wondertan — APPROVED** [review]
Thanks for moving this ❤️

**Comment by Wondertan** [comment]
I lost my nmt approval perms😢

File: hasher.go [hunk 3]
```diff
@@ -71,16 +74,70 @@ func (n *Hasher) NamespaceSize() namespace.IDSize {
...
+// Size returns the number of bytes Sum will return.
+func (n *Hasher) Size() int {
+	return n.baseHasher.Size() + int(n.NamespaceLen)*2
+}
+
+// Write writes the namespaced data to be hashed.
+//
+// Requires data of fixed size to match leaf or inner NMT nodes.
+// Only a single write is allowed.
+func (n *Hasher) Write(data []byte) (int, error) {
```
Comment (adlerjohn) [inline hasher.go:93]: Does this implement another interface? If so, can you document? If it doesn't implement an interface then I'd say just don't allow more than one call to `Write` by construction.

File: nmt.go [hunk 4]
```diff
@@ -23,6 +28,7 @@ type NodeVisitorFn = func(hash []byte, children ...[]byte)
 type Options struct {
 	InitialCapacity    int
 	NamespaceIDSize    namespace.IDSize
+	ShareSize          int
```
Comment (Wondertan) [inline nmt.go:31]: Randomly found a relic that we can remove

File: hasher.go (hunk 3 above)
Comment (Wondertan) [reply hasher.go:93]: It implements `hash.Hash`, which implements `io.Writer`.

**Review by adlerjohn — APPROVED** [review]

**Pull request merged**